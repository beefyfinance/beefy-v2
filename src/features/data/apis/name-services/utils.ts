import { uniq } from 'lodash-es';
import { type Address, getAddress, type Hash, isAddress, numberToHex } from 'viem';
import { namehash, normalize } from 'viem/ens';
import { ZERO_ADDRESS } from '../../../../helpers/addresses.ts';
import type { ChainId } from '../../entities/chain.ts';

export function hashDomain(domain: string): Hash {
  return namehash(domain);
}

export function normalizeDomain(domain: string): string {
  return normalize(domain);
}

export function normalizeAndHashDomain(domain: string): Hash {
  return hashDomain(normalizeDomain(domain));
}

export function normalizeAddress(address: unknown): Address | undefined {
  if (typeof address === 'string' && address.length === 42 && address !== ZERO_ADDRESS) {
    if (isAddress(address, { strict: false })) {
      return getAddress(address);
    }
  }

  return undefined;
}

export function getAllChainsFromTldToChain(tldToChain: Record<string, ChainId[]>): ChainId[] {
  return uniq(Object.values(tldToChain).flat());
}

/**
 * ENSIP-11: EVM compatible Chain Address Resolution
 * @see https://docs.ens.domains/ensip/11/
 **/
export function chainIdToCoinType(chainId: number): number {
  return (0x80000000 | chainId) >>> 0;
}

/**
 * ENSIP-19: Multichain Primary Names
 * @see https://docs.ens.domains/ensip/19/
 */
export function coinTypeToReverseDomain(coinType: number): string {
  switch (coinType) {
    case 60:
      return 'addr.reverse';
    case 0x8000_0000:
      return 'default.reverse';
    default: {
      if (coinType < 0 || coinType > 0xffff_ffff) {
        throw new Error(`Invalid coin type: ${coinType}`);
      }
      return `${numberToHex(coinType, { size: 4 }).slice(2)}.reverse`;
    }
  }
}

/**
 * ENSIP-19: Multichain Primary Names
 * @see https://docs.ens.domains/ensip/19/
 */
export function chainIdToReverseDomain(chainId: number): string {
  return coinTypeToReverseDomain(chainIdToCoinType(chainId));
}
