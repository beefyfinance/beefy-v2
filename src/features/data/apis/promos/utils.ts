export function extractChainId(path: string): string {
  const match = path.match(/.+\/chain\/(.+).json$/);
  const chainId = match?.[1];
  if (chainId) {
    return chainId;
  }
  throw new Error(`Could not extract chain id from path: ${path}`);
}
