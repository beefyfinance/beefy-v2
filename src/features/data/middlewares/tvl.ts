import { isAnyOf } from '@reduxjs/toolkit';
import { BIG_ONE, BIG_ZERO } from '../../../helpers/big-number.ts';
import { fetchAllContractDataByChainAction } from '../actions/contract-data.ts';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../actions/tokens.ts';
import type { FetchAllContractDataResult } from '../apis/contract-data/contract-data-types.ts';
import { isStandardVault, type VaultCowcentrated, type VaultEntity } from '../entities/vault.ts';
import type { BoostTvlById, ChainTvlById, VaultTvlById } from '../reducers/tvl-types.ts';
import { setTvlContractState } from '../reducers/tvl.ts';
import { selectBoostById } from '../selectors/boosts.ts';
import { selectActiveChainIds } from '../selectors/chains.ts';
import { selectTokenByAddress, selectTokenPriceByAddress } from '../selectors/tokens.ts';
import { selectVaultById } from '../selectors/vaults.ts';
import type { BeefyState } from '../store/types.ts';
import { mooAmountToOracleAmount } from '../utils/ppfs.ts';
import { startAppListening } from './listener-middleware.ts';

const contractDataChanged = isAnyOf(
  fetchAllContractDataByChainAction.fulfilled.match,
  reloadBalanceAndAllowanceAndGovRewardsAndBoostData.fulfilled.match
);

export function addTvlListeners() {
  startAppListening({
    matcher: contractDataChanged,
    effect: async (action, { getState, dispatch }) => {
      // TODO this could probably be done in the original actions but we have delayed fulfills in initAppData
      const { contractData } = action.payload;
      const state = getState();
      const byVaultId = getVaultTvlById(state, contractData);
      const byBoostId = getBoostTvlById(state, contractData);
      const { totalTvl, byChaindId } = getTotalTvlByChainId(state, byVaultId);
      dispatch(
        setTvlContractState({
          totalTvl,
          byVaultId,
          byBoostId,
          byChaindId,
        })
      );
    },
  });
}

function getTotalTvlByChainId(state: BeefyState, byVaultId: VaultTvlById) {
  const activeChainIds = selectActiveChainIds(state);
  const byChaindId: ChainTvlById = {};
  let totalTvl = BIG_ZERO;

  for (const [vaultId, vaultTvl] of Object.entries(byVaultId)) {
    const vault = selectVaultById(state, vaultId);

    byChaindId[vault.chainId] = (byChaindId[vault.chainId] || BIG_ZERO).plus(vaultTvl.tvl);

    // Only include active chains in total
    if (activeChainIds.includes(vault.chainId)) {
      totalTvl = totalTvl.plus(vaultTvl.tvl);
    }
  }

  return { totalTvl, byChaindId };
}

function getBoostTvlById(state: BeefyState, contractData: FetchAllContractDataResult) {
  const byBoostId: BoostTvlById = { ...state.biz.tvl.byBoostId };

  // create an index of ppfs for boost tvl usage
  const ppfsPerVaultId: {
    [vaultId: VaultEntity['id']]: BigNumber;
  } = {};
  for (const vault of contractData.standardVaults) {
    ppfsPerVaultId[vault.id] = vault.pricePerFullShare;
  }
  for (const vault of contractData.erc4626Vaults) {
    ppfsPerVaultId[vault.id] = vault.pricePerFullShare;
  }

  /**
   * Boosts always have an associated vault.
   * When computing the apr for a vault + tvl for boost we need to use both vault data and boost data
   * 2 tokens participate in this, one is the token being
   * staked in the vault and the second one is the boost reward token
   * For everything related to totalStaked/tvl we need to work with what's deposited, so the vault's token
   * When talking about apr, we need the totalStaked amount + data from the reward token,
   * since the ratio between the two will give us how much we are earning
   *
   * So let's take as an example
   * if we want to find the boost's tvl to see how much is deposited there
   * since we are talking about tvl we want to know how much has been deposited in that boost
   * so we'd need to know how much is deposited, that would be totalStaked
   * that would give us the amount deposited in the boost
   * however, those can be mooTokens
   * if we have mooTokens, we need to multiply times ppfs to get the amount of the underlying token
   * now that we have the amount of the underlying token
   * we can multiply it times the price and we get the amount  in usd
   */
  for (const boostContractData of contractData.boosts) {
    const boost = selectBoostById(state, boostContractData.id);
    const vault = selectVaultById(state, boost.vaultId);
    const oraclePrice = selectTokenPriceByAddress(state, vault.chainId, vault.depositTokenAddress);
    // find vault price per full share for the vault
    const ppfs = ppfsPerVaultId[vault.id];
    // Only CLM vaults need/have ppfs
    if (ppfs === undefined && isStandardVault(vault)) {
      throw new Error(`Could not find ppfs for vault id ${vault.id}`);
    }
    const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
    const mooToken = selectTokenByAddress(state, vault.chainId, vault.contractAddress);
    const totalStaked = mooAmountToOracleAmount(
      mooToken,
      depositToken,
      ppfs || BIG_ONE,
      boostContractData.totalSupply
    );
    const tvl = totalStaked.times(oraclePrice);

    // add data to state
    byBoostId[boost.id] = { tvl, staked: totalStaked };
  }

  return byBoostId;
}

function getVaultTvlById(state: BeefyState, contractData: FetchAllContractDataResult) {
  const byVaultId: VaultTvlById = { ...state.biz.tvl.byVaultId };

  for (const vaultContractData of contractData.standardVaults) {
    const vault = selectVaultById(state, vaultContractData.id);
    const price = selectTokenPriceByAddress(state, vault.chainId, vault.depositTokenAddress);
    const rawTvl = vaultContractData.balance.times(price);
    const tvl = rawTvl;

    byVaultId[vault.id] = { tvl, rawTvl };
  }

  for (const govVaultContractData of contractData.govVaults) {
    const vault = selectVaultById(state, govVaultContractData.id);
    const price = selectTokenPriceByAddress(state, vault.chainId, vault.depositTokenAddress);
    const totalStaked = govVaultContractData.totalSupply;
    const rawTvl = totalStaked.times(price);
    let tvl = rawTvl;

    // exclude other tvls from counting towards this
    for (const excludedId of vault.excludedIds) {
      const excludedVault = selectVaultById(state, excludedId);
      if (excludedVault && excludedVault.status === 'active') {
        const excludedTVL = byVaultId[excludedId]?.tvl;
        if (excludedTVL) {
          tvl = tvl.minus(excludedTVL);
        }
      }
    }

    byVaultId[vault.id] = { tvl, rawTvl };
  }

  for (const govVaultMultiContractData of contractData.govVaultsMulti) {
    const vault = selectVaultById(state, govVaultMultiContractData.id);
    const price = selectTokenPriceByAddress(state, vault.chainId, vault.depositTokenAddress);
    const totalStaked = govVaultMultiContractData.totalSupply;
    const rawTvl = totalStaked.times(price);
    let tvl = rawTvl;

    // exclude other tvls from counting towards this e.g. bifi gov pool excludes bifi vault standard tvl / clm pool excludes clm vault
    for (const excludedId of vault.excludedIds) {
      const excludedVault = selectVaultById(state, excludedId);
      if (excludedVault && excludedVault.status === 'active') {
        const excludedTVL = byVaultId[excludedId]?.tvl;
        if (excludedTVL) {
          tvl = tvl.minus(excludedTVL);
        }
      }
    }

    byVaultId[vault.id] = { tvl, rawTvl };
  }

  for (const cowVaultContractData of contractData.cowVaults) {
    const vault = selectVaultById(state, cowVaultContractData.id) as VaultCowcentrated;
    const vaultTokens = vault.depositTokenAddresses.map(address =>
      selectTokenByAddress(state, vault.chainId, address)
    );
    const rawTvl = vaultTokens.reduce((acc, token, i) => {
      const price = selectTokenPriceByAddress(state, vault.chainId, token.address);
      return acc.plus(cowVaultContractData.balances[i].times(price));
    }, BIG_ZERO);
    let tvl = rawTvl;

    // exclude other tvls from counting towards this e.g. clm excludes clm pool and clm vault
    for (const excludedId of vault.excludedIds) {
      const excludedVault = selectVaultById(state, excludedId);
      if (excludedVault && excludedVault.status === 'active') {
        const excludedTVL = byVaultId[excludedId]?.tvl;
        if (excludedTVL) {
          tvl = tvl.minus(excludedTVL);
        }
      }
    }

    byVaultId[vault.id] = { tvl, rawTvl };
  }

  for (const erc4626ContractData of contractData.erc4626Vaults) {
    const vault = selectVaultById(state, erc4626ContractData.id);
    const price = selectTokenPriceByAddress(state, vault.chainId, vault.depositTokenAddress);
    const rawTvl = erc4626ContractData.balance.times(price);
    const tvl = rawTvl;

    byVaultId[vault.id] = { tvl, rawTvl };
  }

  return byVaultId;
}
