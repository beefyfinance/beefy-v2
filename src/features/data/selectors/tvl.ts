import { VaultEntity } from '../entities/vault';
import { BeefyState } from '../../../redux-types';
import { BIG_ZERO } from '../../../helpers/big-number';

export const selectVaultTvl = (state: BeefyState, vaultId: VaultEntity['id']) =>
  state.biz.tvl.byVaultId[vaultId]?.tvl || BIG_ZERO;

export const selectTotalTvl = (state: BeefyState) => state.biz.tvl.totalTvl;

export const selectTvlByChain = (state: BeefyState) => state.biz.tvl.byChaindId;
