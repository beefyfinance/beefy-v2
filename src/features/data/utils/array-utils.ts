import { isArray, isPlainObject, mapValues } from 'lodash';

// https://github.com/lodash/lodash/issues/1244#issuecomment-356676695
export const mapValuesDeep = (obj: any, fn: (val: any, key: string, obj: any) => any) =>
  isArray(obj)
    ? obj.map(item => mapValuesDeep(item, fn))
    : mapValues(obj, (val, key) =>
        isPlainObject(val) || isArray(val) ? mapValuesDeep(val, fn) : fn(val, key, obj)
      );

export function createIdMap<T extends { id: string }>(
  arr: T[],
  getId: (item: T) => string = item => item.id
) {
  return arr.reduce((agg, item) => {
    agg[getId(item)] = item;
    return agg;
  }, {} as { [id: string]: T });
}

//https://github.com/lodash/lodash/issues/2339#issuecomment-585615971
export const intersperse = <T>(arr: T[], separator: (n: number) => T): T[] =>
  arr.reduce<T[]>((acc, currentElement, currentIndex) => {
    const isLast = currentIndex === arr.length - 1;
    return [...acc, currentElement, ...(isLast ? [] : [separator(currentIndex)])];
  }, []);
