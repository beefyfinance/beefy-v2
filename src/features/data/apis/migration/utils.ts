import { createAsyncThunk } from '@reduxjs/toolkit';
import type {
  CommonMigrationUpdateFulfilledPayload,
  MigratorExecuteProps,
  MigratorUpdateProps,
} from './migration-types';
import type { BeefyState } from '../../../../redux-types';
import { selectVaultById } from '../../selectors/vaults';
import { selectTokenByAddress } from '../../selectors/tokens';
import { selectUserBalanceToMigrateByVaultId } from '../../selectors/migration';
import type { Step } from '../../reducers/wallet/stepper';
import { walletActions } from '../../actions/wallet-actions';
import { startStepperWithSteps } from '../../actions/stepper';
import { isTokenErc20 } from '../../entities/token';
import { selectAllowanceByTokenAddress } from '../../selectors/allowances';
import type { VaultEntity } from '../../entities/vault';
import BigNumber from 'bignumber.js';
import type Web3 from 'web3';
import { selectChainById } from '../../selectors/chains';
import { getWalletConnectionApi, getWeb3Instance } from '../instances';

export function buildFetchBalance(
  id: string,
  fetchBalance: (
    vault: VaultEntity,
    web3: Web3,
    walletAddress: string,
    state: BeefyState
  ) => Promise<string>
) {
  return createAsyncThunk<
    CommonMigrationUpdateFulfilledPayload,
    MigratorUpdateProps,
    { state: BeefyState }
  >(`migration/${id}/update`, async ({ vaultId, walletAddress }, { getState }) => {
    const state = getState();
    const vault = selectVaultById(state, vaultId);
    const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
    const chain = selectChainById(state, vault.chainId);
    const web3 = await getWeb3Instance(chain);

    const balance = await fetchBalance(vault, web3, walletAddress, state);
    const fixedBalance = new BigNumber(balance).shiftedBy(-depositToken.decimals);
    console.debug(id, vault.id, fixedBalance.toNumber());
    return { vaultId, walletAddress, balance: fixedBalance, migrationId: id };
  });
}

export function buildExecute(
  id: string,
  unstakeCall: (
    vault: VaultEntity,
    web3: Web3,
    amount: BigNumber,
    state: BeefyState
    // eslint-disable-next-line
  ) => Promise<any>
) {
  return createAsyncThunk<void, MigratorExecuteProps, { state: BeefyState }>(
    `migration/${id}/execute`,
    async ({ vaultId, t, migrationId }, { getState, dispatch }) => {
      const steps: Step[] = [];
      const state = getState();
      const vault = selectVaultById(state, vaultId);
      const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
      const walletApi = await getWalletConnectionApi();
      const web3 = await walletApi.getConnectedWeb3Instance();
      const { balance } = selectUserBalanceToMigrateByVaultId(state, vaultId, migrationId);

      const call = await unstakeCall(vault, web3, balance, state);

      steps.push({
        step: 'migration',
        message: t('Vault-MigrationStart'),
        action: walletActions.migrateUnstake(
          call,
          vault,
          balance.shiftedBy(depositToken.decimals),
          migrationId
        ),
        pending: false,
        extraInfo: { vaultId },
      });

      if (isTokenErc20(depositToken)) {
        const allowance = selectAllowanceByTokenAddress(
          state,
          depositToken.chainId,
          depositToken.address,
          vault.earnContractAddress
        );
        if (allowance.lt(balance)) {
          steps.push({
            step: 'approve',
            message: t('Vault-ApproveMsg'),
            action: walletActions.approval(depositToken, vault.earnContractAddress),
            pending: false,
          });
        }
      }

      steps.push({
        step: 'deposit',
        message: t('Vault-TxnConfirm', { type: t('Deposit-noun') }),
        action: walletActions.deposit(vault, balance, true),
        pending: false,
        extraInfo: { vaultId: vault.id },
      });

      dispatch(startStepperWithSteps(steps, vault.chainId));
    }
  );
}
