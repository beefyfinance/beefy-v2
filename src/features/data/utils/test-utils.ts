import { Action } from 'redux';
import { configureStore } from '@reduxjs/toolkit';
import { fetchAllBoosts } from '../actions/boosts';
import { fetchChainConfigs } from '../actions/chains';
import { fetchAllPricesAction } from '../actions/prices';
import { fetchAllVaults } from '../actions/vaults';
import { selectAllChains } from '../selectors/chains';
import mockPrices from './mock-prices.json';
import mockLPPrices from './mock-lp-prices.json';
import { rootReducer } from '../reducers';

/**
 * Create a new BeefyState store with some data included
 * This is used in tests to get a decent starting state
 * when doing things like TVL or APY computation
 * TODO: maybe memoize that, we have to deep freeze the output for that
 */
export async function getBeefyTestingStore() {
  // here, the store has all the proper TS typings
  const store = configureStore({
    reducer: rootReducer,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        // because we use BigNumber which is not serializable by default
        // we disable rerialization altogether
        // a better solution would be to allow serialization of the store
        serializableCheck: false,

        // this makes the old code bug
        immutableCheck: false,
      }),
  });

  let state = store.getState();

  // trigger all configs actions
  await store.dispatch(fetchChainConfigs());
  state = store.getState();
  const chains = selectAllChains(state);
  await store.dispatch(fetchAllVaults({}));
  await store.dispatch(fetchAllBoosts());

  // mock token prices
  store.dispatch({
    type: fetchAllPricesAction.fulfilled,
    payload: { ...mockPrices, ...mockLPPrices },
  });

  return store;
}
