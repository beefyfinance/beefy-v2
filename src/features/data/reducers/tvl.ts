import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import { BIG_ZERO } from '../../../helpers/format';
import { mooAmountToOracleAmount } from '../utils/ppfs';
import { WritableDraft } from 'immer/dist/internal';
import { fetchAllContractDataByChainAction } from '../actions/contract-data';
import { BoostEntity } from '../entities/boost';
import { VaultEntity, VaultGov } from '../entities/vault';
import { selectBoostById } from '../selectors/boosts';
import { selectTokenById, selectTokenPriceByTokenId } from '../selectors/tokens';
import { selectVaultById } from '../selectors/vaults';
import { FetchAllContractDataResult } from '../apis/contract-data/contract-data-types';
import { BeefyState } from '../../../redux-types';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../actions/tokens';

/**
 * State containing APY infos indexed by vault id
 */
export interface TvlState {
  totalTvl: BigNumber;
  byVaultId: {
    [vaultId: VaultEntity['id']]: {
      tvl: BigNumber;
    };
  };
  byBoostId: {
    [boostId: BoostEntity['id']]: {
      staked: BigNumber;
      tvl: BigNumber;
    };
  };
  exclusions: {
    [vaultId: VaultEntity['id']]: VaultEntity['id'];
  };
}
export const initialTvlState: TvlState = {
  totalTvl: BIG_ZERO,
  byVaultId: {},
  byBoostId: {},
  exclusions: {},
};

export const tvlSlice = createSlice({
  name: 'tvl',
  initialState: initialTvlState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    /**
     * On gov vault contract data, recompute tvl and exclusions
     */
    builder.addCase(fetchAllContractDataByChainAction.fulfilled, (sliceState, action) => {
      const state = action.payload.state;
      addContractDataToState(state, sliceState, action.payload.data);
    });

    builder.addCase(
      reloadBalanceAndAllowanceAndGovRewardsAndBoostData.fulfilled,
      (sliceState, action) => {
        const state = action.payload.state;
        addContractDataToState(state, sliceState, action.payload.contractData);
      }
    );
  },
});

function addContractDataToState(
  state: BeefyState,
  sliceState: WritableDraft<TvlState>,
  contractData: FetchAllContractDataResult
) {
  // On standard vault contract data, recompute tvl and exclusions
  for (const vaultContractData of contractData.standardVaults) {
    const vault = selectVaultById(state, vaultContractData.id);
    const price = selectTokenPriceByTokenId(state, vault.oracleId);

    const vaultTvl = vaultContractData.balance.times(price);

    // save for vault
    sliceState.byVaultId[vault.id] = { tvl: vaultTvl };
  }

  for (const govVaultContractData of contractData.govVaults) {
    const totalStaked = govVaultContractData.totalSupply;

    const vault = selectVaultById(state, govVaultContractData.id) as VaultGov;
    const price = selectTokenPriceByTokenId(state, vault.oracleId);

    let tvl = totalStaked.times(price);

    // handle gov vault TVL exclusion
    if (vault.excludedId) {
      const excludedTVL = sliceState.byVaultId[vault.excludedId]?.tvl;
      if (excludedTVL) {
        tvl = tvl.minus(excludedTVL);
      }
    }
    sliceState.byVaultId[vault.id] = { tvl: tvl };
  }

  // create an index of ppfs for boost tvl usage
  const ppfsPerVaultId: { [vaultId: VaultEntity['id']]: BigNumber } = {};
  for (const vault of contractData.standardVaults) {
    ppfsPerVaultId[vault.id] = vault.pricePerFullShare;
  }

  /**
   * Boosts always have an associated vault.
   * When computing the apr for a vault + tvl for boost we need to use both vault data and boost data
   * 2 tokens participate in this, one is the token being
   * staked in the vault and the second one is the boost reward token
   * For everything related to totalStaked/tvl we need to work with what's deposited, so the vault's token
   * When talking about apr, we need the totalStaked amount + data from the reward token,
   * since the ratio between the two will give us how much we are earning
   *
   * So let's take as an example
   * if we want to find the boost's tvl to see how much is deposited there
   * since we are talking about tvl we want to know how much has been deposited in that boost
   * so we'd need to know how much is deposited, that would be totalStaked
   * that would give us the amount deposited in the boost
   * however, those can be mooTokens
   * if we have mooTokens, we need to multiply times ppfs to get the amount of the underlying token
   * now that we have the amount of the underlying token
   * we can multiply it times the price and we get the amount  in usd
   */
  for (const boostContractData of contractData.boosts) {
    const boost = selectBoostById(state, boostContractData.id);
    const vault = selectVaultById(state, boost.vaultId);
    const oraclePrice = selectTokenPriceByTokenId(state, vault.oracleId);
    // find vault price per full share for the vault
    const ppfs = ppfsPerVaultId[vault.id];
    if (ppfs === undefined) {
      throw new Error(`Could not find ppfs for vault id ${vault.id}`);
    }
    const oracleToken = selectTokenById(state, vault.chainId, vault.oracleId);
    const mooToken = selectTokenById(state, vault.chainId, vault.earnedTokenId);
    const totalStaked = mooAmountToOracleAmount(
      mooToken,
      oracleToken,
      ppfs,
      boostContractData.totalSupply
    );
    const tvl = totalStaked.times(oraclePrice);

    // add data to state
    sliceState.byBoostId[boost.id] = { tvl, staked: totalStaked };
  }

  // recompute total tvl as a whole
  let totalTvl = BIG_ZERO;
  for (const vaultTvl of Object.values(sliceState.byVaultId)) {
    totalTvl = totalTvl.plus(vaultTvl.tvl);
  }
  sliceState.totalTvl = totalTvl;
}
