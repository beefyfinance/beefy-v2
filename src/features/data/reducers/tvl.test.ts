import BigNumber from 'bignumber.js';
import {
  fetchAllContractDataByChainAction,
  FetchAllContractDataFulfilledPayload,
} from '../actions/contract-data';
import { getBeefyTestingStore } from '../utils/test-utils';
import { tvlSlice, initialTvlState } from './tvl';

describe('TVL slice tests', () => {
  it('should update state on fulfilled gov vault contract data', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const state = store.getState();

    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        boosts: [],
        standardVaults: [],
        govVaults: [
          {
            id: 'bifi-gov',
            totalSupply: new BigNumber(123).times(18),
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const newState = tvlSlice.reducer(initialTvlState, action);
    expect(newState).toMatchSnapshot();
  });

  // TODO: have a test for testing exclusions

  it('should update state on fulfilled standard vault contract data', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const state = store.getState();

    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        boosts: [],
        govVaults: [],
        standardVaults: [
          {
            id: 'banana-cyt-bnb',
            balance: new BigNumber(123),
            pricePerFullShare: new BigNumber(12),
            strategy: '0x00000000000000000000000',
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const newState = tvlSlice.reducer(initialTvlState, action);
    expect(newState).toMatchSnapshot();
  });

  it('should update state on fulfilled boost contract data', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    let state = store.getState();

    // We want to make sure we handle the action properly
    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        boosts: [
          {
            id: 'moo_banana-banana-busd-bitcrush',
            periodFinish: new Date(2022, 0, 1, 0, 0, 0),
            rewardRate: new BigNumber(0.4),
            totalSupply: new BigNumber(12345),
          },
        ],
        govVaults: [],
        standardVaults: [
          {
            id: 'banana-banana-busd',
            balance: new BigNumber(123),
            pricePerFullShare: new BigNumber(12),
            strategy: '0x00000000000000000000000',
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const newState = tvlSlice.reducer(state.biz.tvl, action);
    expect(newState).toMatchSnapshot();
  });

  it('should fully recompute total TVL when getting new chain data', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    let state = store.getState();

    const bscPayload1: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        boosts: [],
        govVaults: [],
        standardVaults: [
          {
            id: 'banana-cyt-bnb',
            balance: new BigNumber(123),
            pricePerFullShare: new BigNumber(12),
            strategy: '0x00000000000000000000000',
          },
        ],
      },
      state,
    };

    const bscPayload2: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        boosts: [],
        govVaults: [],
        standardVaults: [
          {
            id: 'banana-cyt-bnb',
            balance: new BigNumber(3456),
            pricePerFullShare: new BigNumber(456),
            strategy: '0x00000000000000000000000',
          },
        ],
      },
      state,
    };
    const avaxPayload: FetchAllContractDataFulfilledPayload = {
      chainId: 'avax',
      data: {
        boosts: [],
        govVaults: [],
        standardVaults: [
          {
            id: 'joe-mai-wavax-eol',
            balance: new BigNumber(123),
            pricePerFullShare: new BigNumber(12),
            strategy: '0x00000000000000000000000',
          },
        ],
      },
      state,
    };

    // first, we get some bsc info
    let action = { type: fetchAllContractDataByChainAction.fulfilled, payload: bscPayload1 };
    let sliceState = tvlSlice.reducer(initialTvlState, action);
    expect(sliceState.totalTvl).toMatchSnapshot();

    // then we get avax infos
    action = { type: fetchAllContractDataByChainAction.fulfilled, payload: avaxPayload };
    sliceState = tvlSlice.reducer(sliceState, action);
    expect(sliceState.totalTvl).toMatchSnapshot();

    // then, we get new bsc info, we should not keep adding up but recompute TVL from scratch
    action = { type: fetchAllContractDataByChainAction.fulfilled, payload: bscPayload2 };
    sliceState = tvlSlice.reducer(sliceState, action);
    expect(sliceState.totalTvl).toMatchSnapshot();
  });
});
