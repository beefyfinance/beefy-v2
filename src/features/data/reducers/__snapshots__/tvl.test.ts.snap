// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TVL slice tests should fully recompute total TVL when getting new chain data 1`] = `"817.047632700577131"`;

exports[`TVL slice tests should fully recompute total TVL when getting new chain data 2`] = `"940.047632700577131"`;

exports[`TVL slice tests should fully recompute total TVL when getting new chain data 3`] = `"23080.045679782069632"`;

exports[`TVL slice tests should update state on fulfilled boost contract data 1`] = `
Object {
  "byBoostId": Object {
    "moo_banana-banana-busd-bitcrush": Object {
      "staked": "148140",
      "tvl": "310127.810437199460438",
    },
  },
  "byVaultId": Object {
    "banana-banana-busd": Object {
      "tvl": "257.4977769932194791",
    },
  },
  "exclusions": Object {},
  "totalTvl": "257.4977769932194791",
}
`;

exports[`TVL slice tests should update state on fulfilled gov vault contract data 1`] = `
Object {
  "byBoostId": Object {},
  "byVaultId": Object {
    "bifi-gov": Object {
      "tvl": "3839731.6410978912054",
    },
  },
  "exclusions": Object {},
  "totalTvl": "3839731.6410978912054",
}
`;

exports[`TVL slice tests should update state on fulfilled standard vault contract data 1`] = `
Object {
  "byBoostId": Object {},
  "byVaultId": Object {
    "banana-cyt-bnb": Object {
      "tvl": "817.047632700577131",
    },
  },
  "exclusions": Object {},
  "totalTvl": "817.047632700577131",
}
`;
